virtual_ecosystem/models/animals/animal_cohorts.py

```diff
@@ -195,145 +188,418 @@ def die_individual(self, number_dead: int, carcass_pool: CarcassPool) -> None:
+        self, actual_consumed_mass: float, predator: Consumer, carcass_pool: DecayPool
+    ) -> None:
+        """Updates the carcass pool based on consumed mass and predator's efficiency.
 
-        Note: AnimalCohort mass_current is mean per individual mass within the
-            cohort. Mass is not lost from mass_current from a predation event but the
-            number of individuals in the cohort is reduced.
+        Args:
+            actual_consumed_mass: The total mass consumed from the prey cohort.
+            predator: The predator consuming the cohort.
+            carcass_pool: The pool to which remains of eaten individuals are delivered.
+        """
+        # Adjust the mass for predator's mechanical efficiency to calculate carcass mass
+        carcass_mass = actual_consumed_mass * (
+            1 - predator.functional_group.mechanical_efficiency
+        )
 
-        Currently, this function is in an inbetween state where mass is removed from
-        the animal cohort but it is recieved by the litter pool as energy. This will be
-        fixed once the litter pools are updated for mass.
+        # Update the carcass pool with the remainder
+        carcass_pool.scavengeable_energy += (
+            1 - self.decay_fraction_carcasses
+        ) * carcass_mass
+        carcass_pool.decomposed_energy += self.decay_fraction_carcasses * carcass_mass
 
-        TODO: Rework after update litter pools for mass
+    def get_eaten(
+        self,
+        potential_consumed_mass: float,
+        predator: Consumer,
+        carcass_pool: DecayPool,
+    ) -> float:
+        """Adjusts individuals from the prey cohort based on the consumed mass.
 
         Args:
-            predator: The AnimalCohort preying on the eaten cohort.
-            carcass_pool: The resident pool of animal carcasses to which the remains of
-              dead individuals are delivered.
+            potential_consumed_mass: The mass intended to be consumed by the predator.
+            predator: The predator consuming the cohort.
+            carcass_pool: The pool to which remains of eaten individuals are delivered.
 
         Returns:
-            A float of the mass value of the lost individuals after digestive
-                efficiencies are accounted for.
+            The actual mass consumed by the predator, closely matching consumed_mass.
+        """
+        individual_mass = (
+            self.mass_current
+        )  # Mass of an average individual in the cohort
+        max_individuals_eaten = ceil(potential_consumed_mass / individual_mass)
+        actual_individuals_eaten = min(max_individuals_eaten, self.individuals)
+
+        # Calculate the mass represented by the individuals actually eaten
+        actual_killed_mass = actual_individuals_eaten * individual_mass
+
+        # Calculate the actual amount of mass consumed by the predator
+        actual_consumed_mass = min(actual_killed_mass, potential_consumed_mass)
```

Comment:
I don't think it is the case that `actual_killed_mass <= potential_consumed_mass` is always true, unless I am missing something (unfortunately always possible).

Let's assume:
prey_individuals = 1
prey_mass = 1.0
potential_consumed_mass = 0.5 (this is a possible result from `calculate_consumed_mass_predation` under these conditions)

(I've updated my language a bit here for clarity)
`       # Mass of an average individual in the cohort
        individual_mass = self.mass_current

        max_individuals_killed = ceil(potential_consumed_mass / individual_mass)
        actual_individuals_killed = min(max_individuals_killed, self.individuals)

        # Calculate the mass represented by the individuals actually killed
        actual_mass_killed = actual_individuals_killed * individual_mass

        # Calculate the actual amount of mass consumed by the predator
        actual_mass_consumed = min(actual_mass_killed, potential_consumed_mass)`
        
max_individuals_killed = ceil(0.5 / 1) = 1
actual_individuals_killed = min(1, 1) = 1

actual_mass_killed = 1 * 1.0 = 1.0
actual_mass_consumed = min(1.0, 0.5) = 0.5

In which case, actual_killed_mass is 1.0 and potential_consumed_mass is 0.5. Describing the ecological situation where a predator has killed more than they can actually eat. 

Does that make sense or have I fudged something up?

Reply:
As we discussed on Teams, this makes sense. I'm copying in a comment from that because although it isn't going into this PR, there was some stuff we might want to implement as part of a model review:

> OK - I think the logic makes sense. I think those variable name changes make it clearer.  Another thing I found tricky was the switch in perspectives from predator to prey in those two lines in delta_mass_predation and the naming of the functions.
 
```python
# Calculate the mass to be consumed from this cohort
consumed_mass = self.calculate_consumed_mass_predation(cohort)
# Call get_eaten on the prey cohort to update its mass and individuals
actual_consumed_mass = cohort.get_eaten(consumed_mass, self, carcass_pool)
```
 
> That first value is actually target_mass , right? Nothing has been consumed at all until get_eaten runs. I don't think that target_mass is needed elsewhere and I wonder if having a single function would be more efficient and show the logic more cleanly? I think the same structure is used on plants too?
 
```python
def consume_prey(
    self, target_cohort: AnimalCohort, carcass_pool: DecayPool
) -> float:
    """Consumes prey from a target cohort.
    This method utilizes the F_i_j_individual method to determine the rate at which
    the target cohort is consumed, and then calculates the number of individuals
    killed and the actual mass to be consumed based on this rate and other model
    parameters. The target cohort and carcass pool are also adjusted to reflect the
    predation.
    Args:
        target_cohort: The prey cohort from which mass will be consumed.
        carcass_pool: The pool to which remains of eaten individuals are delivered.
    Returns:
        The actual mass consumed by the predator,(in kg).
    """
    F = self.F_i_j_individual(
        [target_cohort], target_cohort
    )  # Note: Adjusting this call may be necessary
    delta_t = 30.0  # days, TODO: Replace with actual reference or model parameter

    # Calculate the potential consumed mass based on Mad. formula for
    # delta_mass_predation
    potential_consumed_mass = (
        target_cohort.mass_current
        * target_cohort.individuals
        * (
            1
            - exp(-(F * delta_t * self.constants.tau_f * self.constants.sigma_f_t))
        )
    )

    # Calculate the number of individuals killed
    max_individuals_killed = ceil(
        potential_consumed_mass / target_cohort.mass_current
    )
    actual_individuals_killed = min(
        max_individuals_killed, target_cohort.individuals
    )

    # Calculate the mass represented by the individuals killed
    actual_killed_mass = actual_individuals_killed * target_cohort.mass_current

    # Calculate the actual amount of mass consumed by the predator, which can be
    # lower that the killed mass when there is overkill
    actual_consumed_mass = min(actual_killed_mass, potential_consumed_mass)

    # Update the number of individuals in the prey cohort
    target_cohort.individuals -= actual_individuals_killed

    # Calculate the amount of mass that goes into carcass pool and update the pool
    carcass_mass = (actual_killed_mass - actual_consumed_mass) + (
        actual_consumed_mass * (1 - self.functional_group.mechanical_efficiency)
    )

    # Update the carcass pool with carcass mass
    target_cohort.update_carcass_pool(carcass_mass, self, carcass_pool)

    return actual_consumed_mass
```